packages/matter-node-ble.js/src/ble/NobleBleChannel.ts

```diff
@@ -0,0 +1,238 @@
+
+            // So, discover its characteristics.
+            const characteristics = await service.discoverCharacteristicsAsync();
+
+            let characteristicC1ForWrite: Characteristic | undefined;
+            let characteristicC2ForSubscribe: Characteristic | undefined;
+            let additionalCommissioningRelatedData: ByteArray | undefined;
+
+            for (const characteristic of characteristics) {
+                // Loop through each characteristic and match them to the UUIDs that we know about.
+                logger.debug('found characteristic:', characteristic.uuid, characteristic.properties);
+
+                switch (nobleUuidToUuid(characteristic.uuid)) {
+                    case BLE_MATTER_C1_CHARACTERISTIC_UUID:
+                        logger.debug("found C1 characteristic");
+                        characteristicC1ForWrite = characteristic;
+                        break;
+
+                    case BLE_MATTER_C2_CHARACTERISTIC_UUID:
+                        logger.debug("found C2 characteristic");
+                        characteristicC2ForSubscribe = characteristic;
+                        break;
+
+                    case BLE_MATTER_C3_CHARACTERISTIC_UUID:
+                        logger.debug("found C3 characteristic");
+                        if (hasAdditionalAdvertisementData) {
+                            logger.debug("reading additional commissioning related data");
+                            const data = await characteristic.readAsync();
+                            additionalCommissioningRelatedData = new ByteArray(data);
+                        }
+                }
+            }
+
+            if (!characteristicC1ForWrite || !characteristicC2ForSubscribe) {
+                logger.debug('missing characteristics');
+                continue;
+            }
+
+            this.openChannels.set(address, peripheral);
+            return await NobleBleChannel.create(peripheral, characteristicC1ForWrite, characteristicC2ForSubscribe, this.onMatterMessageListener, additionalCommissioningRelatedData);
+        }
+
+        throw new Error(`No Matter service found on peripheral ${peripheral.address}`);
+    }
+
+    onData(listener: (socket: Channel<ByteArray>, data: ByteArray) => void): Listener {
+        this.onMatterMessageListener = listener;
+        return {
+            close: async () => await this.close()
+        }
+    }
+
+    async close() {
+        for (const peripheral of this.openChannels.values()) {
+            await peripheral.disconnectAsync();
+        }
+    }
+}
+
+export class NobleBleChannel implements Channel<ByteArray> {
+
+    static async create(peripheral: Peripheral, characteristicC1ForWrite: Characteristic, characteristicC2ForSubscribe: Characteristic, onMatterMessageListener: ((socket: Channel<ByteArray>, data: ByteArray) => void), _additionalCommissioningRelatedData?: ByteArray): Promise<NobleBleChannel> {
```

Comment:
NBD but rather than async function w/ callback this might be a good candidate for an async generator

Reply:
the callback is not used once, but really as data callback afterwards ...